### PR TITLE
Sites Management Page: Improve render performance by paginating when greater than 96 sites

### LINKED ISSUE #67612: Sites Management Page: Improve perceived performance by applying pagination when > 100 sites are present
When a user has hundreds of sites, simply rendering all of the sites on the Sites page can be quite slow.

[react-virtual-list](https://www.npmjs.com/package/react-virtual-list) is a fancy way to solve this. We may explore it in the future. For now, we can improve perceived performance by simply applying pagination when > 100 sites are present.

From pdKhl6-GX-p2

**Done is:**
* When > 100 sites are present on a given Sites page, we apply the `<Pagination>` component and paginate the list.
* When a user pages through results, a `?page=<number>` query param is added to the URL.
* We also add support for a `?per_page=<number>` hidden query param for modifying the number of results on the page. This will make testing much easier. `per_page` should default to 100.

### PR DESCRIPTION
**Pull request by danielbachhuber**
Fixes #67612

## Proposed Changes

* Applies pagination when displaying greater than 96 sites in order to improve render performance. Uses `96` instead of `100` as the default `perPage` to better support 3 and 2 column rows.
* Exports `handleQueryParamChange` function for use by `<SitesDashboard />` component.

image

## Testing Instructions

1. Navigate to `/sites` with a WordPress user that has greater than 3 sites.
2. Add `?per-page=3` to the URL.
3. Observe the pagination appear at the bottom of the sites list.
4. Click on the pagination component buttons and observe expected behavior (paginating through the results).

One note: when manipulating `?per-page=3`, you may end up in this stacked component scenario:

image

While stacking the pagination and hidden sites components doesn't look like the optimal design, I don't think it's really worth solving for because it's a really uncommon scenario.

### CODE REVIEW

File: client/sites-dashboard/controller.tsx [hunk 1]
```diff
@@ -68,6 +68,10 @@ export function sitesDashboard( context: PageJSContext, next: () => void ) {
 			<PageViewTracker path="/sites" title="Sites Management Page" delay={ 500 } />
 			<SitesDashboard
 				queryParams={ {
+					page: context.query.page ? parseInt( context.query.page ) : undefined,
+					perPage: context.query[ 'per-page' ]
```
Comment (wojtekn) [inline client/sites-dashboard/controller.tsx:72]: Do we want to allow users to set it to any value? Or should we accept the value from the parameter only if it's lower or equal to 96?

**Review by wojtekn — CHANGES_REQUESTED** [review]
@danielbachhuber looks good and works well for me. That feature should help with handling that edge case.

I've just noticed behavior that we may need to adjust - when users are on the second page and start searching for a phrase that can be found on the first page, they will see “No sites match your search”, and they will lose a way to navigate back to the first page. The same happens for the status filter.

Should we reset the page number when users start searching?

There is also a minor styling issue on mobile view - utility buttons partially cover the pagination. We may need to add some more space below the paginator component:

Screen Shot 2022-09-13 at 10 18 56

**Comment by danielbachhuber** [comment]
> I've just noticed behavior that we may need to adjust - when users are on the second page and start searching for a phrase that can be found on the first page, they will see “No sites match your search”, and they will lose a way to navigate back to the first page. The same happens for the status filter.
> 
> Should we reset the page number when users start searching?

@wojtekn Great find! I've updated the logic to reset `page` when a search or filter is initiated: 1699d845ad62bca50b68e36ec475757dc2530418

> There is also a minor styling issue on mobile view - utility buttons partially cover the pagination. We may need to add some more space below the paginator component:

Cleaned up with e8a549f28dbab6794a2c453c0b70187b215434a4 and c67ba827bd840ec71e8298ebe197053b6a2705cb

image

File: client/sites-dashboard/controller.tsx (hunk 1 above)
Comment (danielbachhuber) [reply client/sites-dashboard/controller.tsx:72]: > Do we want to allow users to set it to any value? Or should we accept the value from the parameter only if it's lower or equal to 96?

Given the value only modifies displayed state, and isn't passed through to an API call, I don't think we need to do any more processing than we already are.

**Review by wojtekn — APPROVED** [review]
@danielbachhuber works great now!

**Pull request merged**